README.md
This project is *deprecated* as of 2019-09-14. 

It was mainly used by [OASIS](https://github.com/ocaml/oasis) but the
dependency has been removed in 2017.


CHANGES.txt
2013-11-12 Sylvain Le Gall <syvlain .A.T. le-gall.net>

  * New version 0.0.11
  * Port test suite to OUnit 2.0.0
  * Fix handling of long record that may split at the end of the line with a
    blank.

2012-10-06 Sylvain Le Gall <syvlain .A.T. le-gall.net>

  * New version 0.0.10
  * Use type_conv 108.07.01 (Closes: #1226)

2012-05-23 Sylvain Le Gall <sylvain .A.T. le-gall.net>

  * New version 0.0.9
  * Use type_conv 3.0.5

2012-03-20 Sylvain Le Gall <sylvain .A.T. le-gall.net>

  * New version to 0.0.8
  * Regenerate using oasis 0.3.0~rc3

2011-12-21 Sylvain Le Gall <sylvain .A.T. le-gall.net>

  * New version 0.0.7
  * Use type-conv 3.0.4

2011-06-23 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * New version 0.0.6
 * Fix 'type foo = bar list' which emits 'and odn_of_foo = ODN.of_list
   odn_of_bar', which is not a valid RHS for a letrec
   (thanks to Michael Ekstrand)

2011-06-07 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * New version 0.0.5
 * Upgrade to type-conv 2.3.0 (thanks to a patch provided by Till Varoquaux)

2011-05-24 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * New version 0.0.4
 * Handle polymorphic variants with arguments

2010-10-22 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * New version 0.0.3
 * Limited support for polymorphic variant

2010-09-01 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * New version 0.0.2
 * Compile with OCaml 3.12.0 (Closes: #717)

2010-04-08 Sylvain Le Gall <sylvain.le-gall .A.T. ocamlcore.com>

 * Initial version 0.0.1



test/test.ml
(******************************************************************************)
(* ocaml-data-notation: Store data using OCaml notation                       *)
(*                                                                            *)
(* Copyright (C) 2009-2011, OCamlCore SARL                                    *)
(* Copyright (C) 2013, Sylvain Le Gall                                        *)
(*                                                                            *)
(* This library is free software; you can redistribute it and/or modify it    *)
(* under the terms of the GNU Lesser General Public License as published by   *)
(* the Free Software Foundation; either version 2.1 of the License, or (at    *)
(* your option) any later version, with the OCaml static compilation          *)
(* exception.                                                                 *)
(*                                                                            *)
(* This library is distributed in the hope that it will be useful, but        *)
(* WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *)
(* or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more         *)
(* details.                                                                   *)
(*                                                                            *)
(* You should have received a copy of the GNU Lesser General Public License   *)
(* along with this library; if not, write to the Free Software Foundation,    *)
(* Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA              *)
(******************************************************************************)

(** Main for tests
    @author Sylvain Le Gall
  *)

open OUnit2
open TestCommon
open FileUtil
open ODN

let camlp4o = Conf.make_exec "camlp4o"
let ocamlfind = Conf.make_exec "ocamlfind"

let odn_path_default =
  Filename.concat (Sys.getcwd ()) (Filename.concat "_build" "src")

let odn_path =
  Conf.make_string "odn_path" odn_path_default
    "Directory containing pa_odn.cma and pa_noodn.cma."

let pa_type_conv_cma =
 Filename.concat TestConf.pkg_type_conv_dir "pa_type_conv.cma"

let test_files (dir, fns) =
  let nm =
    match fns with
      | [fn] ->
          dir^"/"^fn
      | lst ->
          dir^"/("^(String.concat "|" fns)^")"
  in
    nm >::
    (fun test_ctxt ->
       let odn_path = odn_path test_ctxt in
       let pa_odn_cma = Filename.concat odn_path "pa_odn.cma" in
       let tmpdn = bracket_tmpdir test_ctxt in
       let () =
         (* Copy files to temporary directory. *)
         FileUtil.cp
           (List.map (fun fn -> in_testdata_dir test_ctxt [dir; fn]) fns)
           tmpdn
       in
       let fns' =
         List.map (Filename.concat tmpdn) fns
       in
       let target_program = Filename.concat tmpdn "test" in
       List.iter
         (fun fn ->
            assert_command ~ctxt:test_ctxt
              (camlp4o test_ctxt)
              [pa_type_conv_cma; pa_odn_cma;
               "Camlp4OCamlPrinter.cmo"; fn];
            assert_command ~ctxt:test_ctxt
              (camlp4o test_ctxt)
              [pa_type_conv_cma; pa_odn_cma;
               "Camlp4AstLifter.cmo"; "Camlp4OCamlPrinter.cmo";
               fn])
         fns';

         assert_command ~ctxt:test_ctxt
           (ocamlfind test_ctxt)
           (["ocamlc"; "-g"; "-o"; target_program;
             "-I"; odn_path; "-I"; tmpdn;
             "-package"; "type_conv"; "-syntax";  "camlp4o";
             "-ppopt"; pa_odn_cma; "odn.cma"] @ fns'))

let () =
  run_test_tt_main
    ("odn">:::
     [
       "pure-odn" >::
       (fun test_ctxt ->
          assert_equal
            ~msg:"variant-simple"
            "Test"
            (string_of_odn (VRT("Test", [])));
          assert_equal
            ~msg:"variant-module"
            "Test"
            (string_of_odn
               ~opened_modules:["MyTest"]
               (VRT("MyTest.Test", [])));
          assert_equal
            ~msg:"long-record"
            ~printer:(fun s -> s)
            "{foo = 123456; bar = 123456; baz = 123456; foo1 = 123456; \
              foo2 = 1; foo3 = 1}"
            (string_of_odn
               ~opened_modules:["Bar"]
               (REC("Bar",
                    [
                      "foo", INT 123456;
                      "bar", INT 123456;
                      "baz", INT 123456;
                      "foo1", INT 123456;
                      "foo2", INT 1;
                      "foo3", INT 1;
                    ])));
       ());

       test_files
         ("oasis-examples",
          ["PropList.ml"; "OASISTypes.ml"; "main.ml"]);

       test_files ("", ["tuples.ml"]);

       "oasis-example no odn" >::
       (fun test_ctxt  ->
          let regexps =
            [
              Str.regexp_string "TYPE_CONV_PATH";
              Str.regexp "with  *odn";
            ]
          in
          let assert_not_match regexp str =
            try
              let _i : int =
                Str.search_forward regexp str 0
              in
                assert_failure
                  (Printf.sprintf
                     "Found '%s' in string '%s'"
                     (Str.matched_string str)
                     str)
            with Not_found ->
              ()
          in
          let lst = ref [] in
          let foutput strm =
            let buff = Buffer.create 13 in
            (* Check that output doesn't contain regexp defined
             * before
             *)
            let flush_buffer () =
              lst := Buffer.contents buff :: !lst;
              Buffer.clear buff
            in
              Stream.iter
                (function
                   | '\n' -> flush_buffer ()
                   | c -> Buffer.add_char buff c)
                strm;
              flush_buffer ()
          in
          let pa_noodn_cma =
            Filename.concat (odn_path test_ctxt) "pa_noodn.cma"
          in

            (* Create a file without odn in it *)
            assert_command ~ctxt:test_ctxt ~foutput
              (camlp4o test_ctxt)
              [pa_type_conv_cma; pa_noodn_cma;
               "Camlp4OCamlPrinter.cmo";
               (in_testdata_dir test_ctxt ["oasis-examples"; "OASISTypes.ml"])];

            List.iter
              (fun line ->
                 List.iter
                  (fun rgxp -> assert_not_match rgxp line)
                  regexps)
              (List.rev !lst)
       );

       test_files ("", ["polyvariants.ml"]);
     ])


tests/test.ml
(******************************************************************************)
(* ocaml-data-notation: Store data using OCaml notation                       *)
(*                                                                            *)
(* Copyright (C) 2009-2011, OCamlCore SARL                                    *)
(* Copyright (C) 2013, Sylvain Le Gall                                        *)
(*                                                                            *)
(* This library is free software; you can redistribute it and/or modify it    *)
(* under the terms of the GNU Lesser General Public License as published by   *)
(* the Free Software Foundation; either version 2.1 of the License, or (at    *)
(* your option) any later version, with the OCaml static compilation          *)
(* exception.                                                                 *)
(*                                                                            *)
(* This library is distributed in the hope that it will be useful, but        *)
(* WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *)
(* or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more         *)
(* details.                                                                   *)
(*                                                                            *)
(* You should have received a copy of the GNU Lesser General Public License   *)
(* along with this library; if not, write to the Free Software Foundation,    *)
(* Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA              *)
(******************************************************************************)

(** Main for tests

    @author Sylvain Le Gall
  *)

open OUnit;;
open TestCommon;;
open FileUtil;;
open ODN;;

let dbug =
  ref false

let _res: test_result list =
  let odn_path =
    Filename.concat (Sys.getcwd ()) (Filename.concat "_build" "src")
  in
  let pa_odn_cma =
    Filename.concat odn_path "pa_odn.cma"
  in
  let pa_noodn_cma =
    Filename.concat odn_path "pa_noodn.cma"
  in
  let pa_type_conv_cma =
    Filename.concat TestConf.pkg_type_conv_dir "pa_type_conv.cma"
  in
  let test_files (dir, fns) =
    let nm =
      match fns with
        | [fn] ->
            dir^"/"^fn
        | lst ->
            dir^"/("^(String.concat "|" fns)^")"
    in
      nm >::
      (bracket
         (fun () ->
            let pwd =
              Sys.getcwd ()
            in
              Sys.chdir dir;
              pwd)
         (fun _ ->
            if !dbug then
              begin
                List.iter
                  (fun fn ->
                     print_endline ("File "^fn^": ");
                     assert_command
                       "camlp4o"
                       [pa_type_conv_cma; pa_odn_cma;
                        "Camlp4OCamlPrinter.cmo"; fn];
                     assert_command
                       "camlp4o"
                       [pa_type_conv_cma; pa_odn_cma;
                        "Camlp4AstLifter.cmo"; "Camlp4OCamlPrinter.cmo";
                        fn])
                  fns
              end;

            assert_command
              "ocamlfind"
              (["ocamlc"; "-g"; "-o"; "test"; "-I"; odn_path;
                "-package"; "type-conv"; "-syntax";  "camlp4o";
                "-ppopt"; pa_odn_cma; "odn.cma"] @ fns)))
         (fun old_cwd ->
            rm
              (filter
                 (Or
                    (Has_extension "cmi",
                     Has_extension "cmo"))
                 (ls "."));
            rm ["test"];
            Sys.chdir old_cwd);
  in
    run_test_tt_main
      ~set_verbose:(fun b -> dbug := b)
      ("odn">:::
       [
         "pure-odn" >::
         (fun () ->
            assert_equal
              ~msg:"variant-simple"
              "Test"
              (string_of_odn (VRT("Test", [])));
            assert_equal
              ~msg:"variant-module"
              "Test"
              (string_of_odn
                 ~opened_modules:["MyTest"]
                 (VRT("MyTest.Test", []))));

         test_files
           ("tests/data/oasis-examples",
            ["PropList.ml"; "OASISTypes.ml"; "main.ml"]);

         test_files
           ("tests/data/",
            ["tuples.ml"]);

         "oasis-example no odn" >::
         (fun () ->
            let regexps =
              [
                Str.regexp_string "TYPE_CONV_PATH";
                Str.regexp "with  *odn";
              ]
            in
            let assert_not_match regexp str =
              try
                let _i : int =
                  Str.search_forward regexp str 0
                in
                  assert_failure
                    (Printf.sprintf
                       "Found '%s' in string '%s'"
                       (Str.matched_string str)
                       str)
              with Not_found ->
                ()
            in
            let foutput strm =
              (* Check that output doesn't contain regexp defined
               * before
               *)
              let buff =
                Buffer.create 13
              in
              let check_buffer () =
                let ln =
                  Buffer.contents buff
                in
                  List.iter
                    (fun rgxp -> assert_not_match rgxp ln)
                    regexps;
                  Buffer.clear buff
              in
                Stream.iter
                  (function
                     | '\n' ->
                         check_buffer ()
                     | c ->
                         Buffer.add_char buff c)
                  strm;
                check_buffer ()
            in

              (* Create a file without odn in it *)
              assert_command
                ~foutput
                "camlp4o"
                [pa_type_conv_cma; pa_noodn_cma;
                 "Camlp4OCamlPrinter.cmo";
                 "tests/data/oasis-examples/OASISTypes.ml"]
         );

         test_files
            ("tests/data/",
             ["polyvariants.ml"]);
       ])
;;


test/TestCommon.ml
(******************************************************************************)
(* ocaml-data-notation: Store data using OCaml notation                       *)
(*                                                                            *)
(* Copyright (C) 2009-2011, OCamlCore SARL                                    *)
(* Copyright (C) 2013, Sylvain Le Gall                                        *)
(*                                                                            *)
(* This library is free software; you can redistribute it and/or modify it    *)
(* under the terms of the GNU Lesser General Public License as published by   *)
(* the Free Software Foundation; either version 2.1 of the License, or (at    *)
(* your option) any later version, with the OCaml static compilation          *)
(* exception.                                                                 *)
(*                                                                            *)
(* This library is distributed in the hope that it will be useful, but        *)
(* WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *)
(* or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more         *)
(* details.                                                                   *)
(*                                                                            *)
(* You should have received a copy of the GNU Lesser General Public License   *)
(* along with this library; if not, write to the Free Software Foundation,    *)
(* Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA              *)
(******************************************************************************)

(** Common functions for tests

    @author Sylvain Le Gall
  *)




setup.ml
(********************************************************************************)
(*  ODN: Dump data using OCaml notation                                         *)
(*                                                                              *)
(*  Copyright (C) 2009-2011, OCamlCore SARL                                     *)
(*                                                                              *)
(*  This library is free software; you can redistribute it and/or modify it     *)
(*  under the terms of the GNU Lesser General Public License as published by    *)
(*  the Free Software Foundation; either version 2.1 of the License, or (at     *)
(*  your option) any later version, with the OCaml static compilation           *)
(*  exception.                                                                  *)
(*                                                                              *)
(*  This library is distributed in the hope that it will be useful, but         *)
(*  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY  *)
(*  or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more          *)
(*  details.                                                                    *)
(*                                                                              *)
(*  You should have received a copy of the GNU Lesser General Public License    *)
(*  along with this library; if not, write to the Free Software Foundation,     *)
(*  Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA               *)
(********************************************************************************)

(* OASIS_START *)
(* DO NOT EDIT (digest: 7f47a529f70709161149c201ccd90f0b) *)
#use "topfind";;
#require "oasis.dynrun";;
open OASISDynRun;;
(* OASIS_STOP *)
let () = setup ();;


src/pa_noodn.ml
(******************************************************************************)
(* ocaml-data-notation: Store data using OCaml notation                       *)
(*                                                                            *)
(* Copyright (C) 2009-2011, OCamlCore SARL                                    *)
(* Copyright (C) 2013, Sylvain Le Gall                                        *)
(*                                                                            *)
(* This library is free software; you can redistribute it and/or modify it    *)
(* under the terms of the GNU Lesser General Public License as published by   *)
(* the Free Software Foundation; either version 2.1 of the License, or (at    *)
(* your option) any later version, with the OCaml static compilation          *)
(* exception.                                                                 *)
(*                                                                            *)
(* This library is distributed in the hope that it will be useful, but        *)
(* WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *)
(* or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more         *)
(* details.                                                                   *)
(*                                                                            *)
(* You should have received a copy of the GNU Lesser General Public License   *)
(* along with this library; if not, write to the Free Software Foundation,    *)
(* Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA              *)
(******************************************************************************)

(** Syntax extension that removes "with odn"
    from files.

  @author Sylvain Le Gall
  *)

open Camlp4
open PreCast
open Ast
open Pa_type_conv

add_generator
  "odn"
  (fun _ _ ->
     let _loc =
       Loc.ghost
     in
       <:str_item<>>)


myocamlbuild.ml
(********************************************************************************)
(*  ODN: Dump data using OCaml notation                                         *)
(*                                                                              *)
(*  Copyright (C) 2009-2011, OCamlCore SARL                                     *)
(*                                                                              *)
(*  This library is free software; you can redistribute it and/or modify it     *)
(*  under the terms of the GNU Lesser General Public License as published by    *)
(*  the Free Software Foundation; either version 2.1 of the License, or (at     *)
(*  your option) any later version, with the OCaml static compilation           *)
(*  exception.                                                                  *)
(*                                                                              *)
(*  This library is distributed in the hope that it will be useful, but         *)
(*  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY  *)
(*  or FITNESS FOR A PARTICULAR PURPOSE. See the file COPYING for more          *)
(*  details.                                                                    *)
(*                                                                              *)
(*  You should have received a copy of the GNU Lesser General Public License    *)
(*  along with this library; if not, write to the Free Software Foundation,     *)
(*  Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA               *)
(********************************************************************************)

(* OASIS_START *)
(* DO NOT EDIT (digest: f922d43ab18f38eb9e477cfc08ed7f6a) *)
module OASISGettext = struct
# 21 "/home/gildor/programmation/oasis/src/oasis/OASISGettext.ml"

  let ns_ str =
    str

  let s_ str =
    str

  let f_ (str : ('a, 'b, 'c, 'd) format4) =
    str

  let fn_ fmt1 fmt2 n =
    if n = 1 then
      fmt1^^""
    else
      fmt2^^""

  let init =
    []

end

module OASISExpr = struct
# 21 "/home/gildor/programmation/oasis/src/oasis/OASISExpr.ml"



  open OASISGettext

  type test = string 

  type flag = string 

  type t =
    | EBool of bool
    | ENot of t
    | EAnd of t * t
    | EOr of t * t
    | EFlag of flag
    | ETest of test * string
    

  type 'a choices = (t * 'a) list 

  let eval var_get t =
    let rec eval' =
      function
        | EBool b ->
            b

        | ENot e ->
            not (eval' e)

        | EAnd (e1, e2) ->
            (eval' e1) && (eval' e2)

        | EOr (e1, e2) ->
            (eval' e1) || (eval' e2)

        | EFlag nm ->
            let v =
              var_get nm
            in
              assert(v = "true" || v = "false");
              (v = "true")

        | ETest (nm, vl) ->
            let v =
              var_get nm
            in
              (v = vl)
    in
      eval' t

  let choose ?printer ?name var_get lst =
    let rec choose_aux =
      function
        | (cond, vl) :: tl ->
            if eval var_get cond then
              vl
            else
              choose_aux tl
        | [] ->
            let str_lst =
              if lst = [] then
                s_ "<empty>"
              else
                String.concat
                  (s_ ", ")
                  (List.map
                     (fun (cond, vl) ->
                        match printer with
                          | Some p -> p vl
                          | None -> s_ "<no printer>")
                     lst)
            in
              match name with
                | Some nm ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for the choice list '%s': %s")
                         nm str_lst)
                | None ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for a choice list: %s")
                         str_lst)
    in
      choose_aux (List.rev lst)

end


# 117 "myocamlbuild.ml"
module BaseEnvLight = struct
# 21 "/home/gildor/programmation/oasis/src/base/BaseEnvLight.ml"

  module MapString = Map.Make(String)

  type t = string MapString.t

  let default_filename =
    Filename.concat
      (Sys.getcwd ())
      "setup.data"

  let load ?(allow_empty=false) ?(filename=default_filename) () =
    if Sys.file_exists filename then
      begin
        let chn =
          open_in_bin filename
        in
        let st =
          Stream.of_channel chn
        in
        let line =
          ref 1
        in
        let st_line =
          Stream.from
            (fun _ ->
               try
                 match Stream.next st with
                   | '\n' -> incr line; Some '\n'
                   | c -> Some c
               with Stream.Failure -> None)
        in
        let lexer =
          Genlex.make_lexer ["="] st_line
        in
        let rec read_file mp =
          match Stream.npeek 3 lexer with
            | [Genlex.Ident nm; Genlex.Kwd "="; Genlex.String value] ->
                Stream.junk lexer;
                Stream.junk lexer;
                Stream.junk lexer;
                read_file (MapString.add nm value mp)
            | [] ->
                mp
            | _ ->
                failwith
                  (Printf.sprintf
                     "Malformed data file '%s' line %d"
                     filename !line)
        in
        let mp =
          read_file MapString.empty
        in
          close_in chn;
          mp
      end
    else if allow_empty then
      begin
        MapString.empty
      end
    else
      begin
        failwith
          (Printf.sprintf
             "Unable to load environment, the file '%s' doesn't exist."
             filename)
      end

  let var_get name env =
    let rec var_expand str =
      let buff =
        Buffer.create ((String.length str) * 2)
      in
        Buffer.add_substitute
          buff
          (fun var ->
             try
               var_expand (MapString.find var env)
             with Not_found ->
               failwith
                 (Printf.sprintf
                    "No variable %s defined when trying to expand %S."
                    var
                    str))
          str;
        Buffer.contents buff
    in
      var_expand (MapString.find name env)

  let var_choose lst env =
    OASISExpr.choose
      (fun nm -> var_get nm env)
      lst
end


# 215 "myocamlbuild.ml"
module MyOCamlbuildFindlib = struct
# 21 "/home/gildor/programmation/oasis/src/plugins/ocamlbuild/MyOCamlbuildFindlib.ml"

  (** OCamlbuild extension, copied from 
    * http://brion.inria.fr/gallium/index.php/Using_ocamlfind_with_ocamlbuild
    * by N. Pouillard and others
    *
    * Updated on 2009/02/28
    *
    * Modified by Sylvain Le Gall 
    *)
  open Ocamlbuild_plugin

  (* these functions are not really officially exported *)
  let run_and_read = 
    Ocamlbuild_pack.My_unix.run_and_read

  let blank_sep_strings = 
    Ocamlbuild_pack.Lexers.blank_sep_strings

  let split s ch =
    let x = 
      ref [] 
    in
    let rec go s =
      let pos = 
        String.index s ch 
      in
        x := (String.before s pos)::!x;
        go (String.after s (pos + 1))
    in
      try
        go s
      with Not_found -> !x

  let split_nl s = split s '\n'

  let before_space s =
    try
      String.before s (String.index s ' ')
    with Not_found -> s

  (* this lists all supported packages *)
  let find_packages () =
    List.map before_space (split_nl & run_and_read "ocamlfind list")

  (* this is supposed to list available syntaxes, but I don't know how to do it. *)
  let find_syntaxes () = ["camlp4o"; "camlp4r"]

  (* ocamlfind command *)
  let ocamlfind x = S[A"ocamlfind"; x]

  let dispatch =
    function
      | Before_options ->
          (* by using Before_options one let command line options have an higher priority *)
          (* on the contrary using After_options will guarantee to have the higher priority *)
          (* override default commands by ocamlfind ones *)
          Options.ocamlc     := ocamlfind & A"ocamlc";
          Options.ocamlopt   := ocamlfind & A"ocamlopt";
          Options.ocamldep   := ocamlfind & A"ocamldep";
          Options.ocamldoc   := ocamlfind & A"ocamldoc";
          Options.ocamlmktop := ocamlfind & A"ocamlmktop"
                                  
      | After_rules ->
          
          (* When one link an OCaml library/binary/package, one should use -linkpkg *)
          flag ["ocaml"; "link"; "program"] & A"-linkpkg";
          
          (* For each ocamlfind package one inject the -package option when
           * compiling, computing dependencies, generating documentation and
           * linking. *)
          List.iter 
            begin fun pkg ->
              flag ["ocaml"; "compile";  "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "ocamldep"; "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "doc";      "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "link";     "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "infer_interface"; "pkg_"^pkg] & S[A"-package"; A pkg];
            end 
            (find_packages ());

          (* Like -package but for extensions syntax. Morover -syntax is useless
           * when linking. *)
          List.iter begin fun syntax ->
          flag ["ocaml"; "compile";  "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "ocamldep"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "doc";      "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "infer_interface"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          end (find_syntaxes ());

          (* The default "thread" tag is not compatible with ocamlfind.
           * Indeed, the default rules add the "threads.cma" or "threads.cmxa"
           * options when using this tag. When using the "-linkpkg" option with
           * ocamlfind, this module will then be added twice on the command line.
           *                        
           * To solve this, one approach is to add the "-thread" option when using
           * the "threads" package using the previous plugin.
           *)
          flag ["ocaml"; "pkg_threads"; "compile"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "doc"] (S[A "-I"; A "+threads"]);
          flag ["ocaml"; "pkg_threads"; "link"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "infer_interface"] (S[A "-thread"])

      | _ -> 
          ()

end

module MyOCamlbuildBase = struct
# 21 "/home/gildor/programmation/oasis/src/plugins/ocamlbuild/MyOCamlbuildBase.ml"

  (** Base functions for writing myocamlbuild.ml
      @author Sylvain Le Gall
    *)



  open Ocamlbuild_plugin
  module OC = Ocamlbuild_pack.Ocaml_compiler

  type dir = string 
  type file = string 
  type name = string 
  type tag = string 

# 56 "/home/gildor/programmation/oasis/src/plugins/ocamlbuild/MyOCamlbuildBase.ml"

  type t =
      {
        lib_ocaml: (name * dir list) list;
        lib_c:     (name * dir * file list) list; 
        flags:     (tag list * (spec OASISExpr.choices)) list;
        (* Replace the 'dir: include' from _tags by a precise interdepends in
         * directory.
         *)
        includes:  (dir * dir list) list; 
      } 

  let env_filename =
    Pathname.basename 
      BaseEnvLight.default_filename

  let dispatch_combine lst =
    fun e ->
      List.iter 
        (fun dispatch -> dispatch e)
        lst 

  let tag_libstubs nm =
    "use_lib"^nm^"_stubs"

  let nm_libstubs nm =
    nm^"_stubs"

  let dispatch t e = 
    let env = 
      BaseEnvLight.load 
        ~filename:env_filename 
        ~allow_empty:true
        ()
    in
      match e with 
        | Before_options ->
            let no_trailing_dot s =
              if String.length s >= 1 && s.[0] = '.' then
                String.sub s 1 ((String.length s) - 1)
              else
                s
            in
              List.iter
                (fun (opt, var) ->
                   try 
                     opt := no_trailing_dot (BaseEnvLight.var_get var env)
                   with Not_found ->
                     Printf.eprintf "W: Cannot get variable %s" var)
                [
                  Options.ext_obj, "ext_obj";
                  Options.ext_lib, "ext_lib";
                  Options.ext_dll, "ext_dll";
                ]

        | After_rules -> 
            (* Declare OCaml libraries *)
            List.iter 
              (function
                 | nm, [] ->
                     ocaml_lib nm
                 | nm, dir :: tl ->
                     ocaml_lib ~dir:dir (dir^"/"^nm);
                     List.iter 
                       (fun dir -> 
                          List.iter
                            (fun str ->
                               flag ["ocaml"; "use_"^nm; str] (S[A"-I"; P dir]))
                            ["compile"; "infer_interface"; "doc"])
                       tl)
              t.lib_ocaml;

            (* Declare directories dependencies, replace "include" in _tags. *)
            List.iter 
              (fun (dir, include_dirs) ->
                 Pathname.define_context dir include_dirs)
              t.includes;

            (* Declare C libraries *)
            List.iter
              (fun (lib, dir, headers) ->
                   (* Handle C part of library *)
                   flag ["link"; "library"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("-l"^(nm_libstubs lib)); A"-cclib";
                        A("-l"^(nm_libstubs lib))]);

                   flag ["link"; "library"; "ocaml"; "native"; tag_libstubs lib]
                     (S[A"-cclib"; A("-l"^(nm_libstubs lib))]);
                        
                   flag ["link"; "program"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("dll"^(nm_libstubs lib))]);

                   (* When ocaml link something that use the C library, then one
                      need that file to be up to date.
                    *)
                   dep ["link"; "ocaml"; "program"; tag_libstubs lib]
                     [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

                   dep  ["compile"; "ocaml"; "program"; tag_libstubs lib]
                     [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

                   (* TODO: be more specific about what depends on headers *)
                   (* Depends on .h files *)
                   dep ["compile"; "c"] 
                     headers;

                   (* Setup search path for lib *)
                   flag ["link"; "ocaml"; "use_"^lib] 
                     (S[A"-I"; P(dir)]);
              )
              t.lib_c;

              (* Add flags *)
              List.iter
              (fun (tags, cond_specs) ->
                 let spec = 
                   BaseEnvLight.var_choose cond_specs env
                 in
                   flag tags & spec)
              t.flags
        | _ -> 
            ()

  let dispatch_default t =
    dispatch_combine 
      [
        dispatch t;
        MyOCamlbuildFindlib.dispatch;
      ]

end


# 476 "myocamlbuild.ml"
open Ocamlbuild_plugin;;
let package_default =
  {
     MyOCamlbuildBase.lib_ocaml =
       [("odn", ["src"]); ("pa_odn", ["src"]); ("pa_noodn", ["src"])];
     lib_c = [];
     flags = [];
     includes = [("tests", ["src"])];
     }
  ;;

let dispatch_default = MyOCamlbuildBase.dispatch_default package_default;;

# 491 "myocamlbuild.ml"
(* OASIS_STOP *)
Ocamlbuild_plugin.dispatch dispatch_default;;
